### PR TITLE
Use upstream MLIR Transform dialect

### PR DESCRIPTION
**Pull request by ftynse**
(depends on #9056)

MLIR upstream now includes the Transform dialect along with op
interfaces and the new extension mechanism. Port most transform ops to
use the upstream interface. The LinalgTransform dialect still contains
the "util" ops for scoped transformation, which arguably should have
been placed in a separate dialect to start with. These will be cleaned
up separately as a decision is required whether to keep them at all.

Notable changes in the naming scheme:
  - `iree_linalg_transform.` goes away, general `transform.` should
    be used instead as the ops are now injected into the transform
    dialect;
  - `*.iree_*` transform ops are now `transform.iree.*` to align on the
    naming scheme of the upstream transform dialect;
  - transform ops related to the structured ops philosophy are prefixed
    with `structured.`, e.g., `transform.structured.pad`;
  - due to a bug in the upstream parser, three-piece op names cannot
    omit the dialect prefix inside transform ops, i.e.
    `transform.structured.pad` must be spelled out completely instead of
    using the shorter `structured.pad` form.

### CODE REVIEW

**Review by MaheshRavishankar — COMMENTED** [review]
Thanks Alex. Looking through this, it seems mostly mechanical move (still a lot of changes though, so thanks for plumbing through it). Adding Nicolas to review this in more detail, but could you just confirm that this is mostly mechanical?

**Comment by ftynse** [comment]
Yes, it is _mostly_ mechanical. There are some changes with respect to have nested transform operations are handled due to the differences between the prototype dialect implementation previously used here and how it ended up being implemented upstream, and the naming scheme changes described in the commit message.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformOpTraits.h [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                  "expected zero- or single-result op");
+    return success();
+  }
+};
+
+template <typename OpTy>
+class FunctionalStyleTransformOpTrait
+    : public OpTrait::TraitBase<OpTy, FunctionalStyleTransformOpTrait> {
+public:
+  void getEffects(SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
```
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformOpTraits.h:126]: This is to allow running CSE and other things on the transform dialect itself, right?
Can you comment why you need this here?

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp [hunk 2]
```diff
@@ -0,0 +1,1296 @@
...
+#define GET_OP_LIST
+#include "iree-dialects/Dialect/LinalgTransform/StructuredTransformOpsExt.cpp.inc"
+      >();
+
+  registerPDLMatchConstraintFn("nestedInFunc", nestedInFunc);
+
+  declareDependentDialect<bufferization::BufferizationDialect>();
+  declareDependentDialect<vector::VectorDialect>();
+
+  // declareDependentDialect<arith::ArithmeticDialect>();
```
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp:98]: dead code

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp [hunk 3]
```diff
@@ -0,0 +1,1296 @@
...
+};
+} // namespace detail
+
+//===----------------------------------------------------------------------===//
+// CanonicalizedSequenceOp
+//===----------------------------------------------------------------------===//
+
+/// Run enabling transformations (LICM and its variants, single-iteration loop
+/// removal, CSE) on the given function.
+static LogicalResult performEnablerTransformations(
```
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp:271]: nit Enabling

**Review by nicolasvasilache — APPROVED** [review]
Looks good, thanks!

**Pull request merged**